File: conformance/code/individual/src/main/java/com/digit/individual/client/VaultClient.java
package com.digit.individual.client;

import com.digit.individual.config.IndividualProperties;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;
import org.springframework.stereotype.Component;
import tools.jackson.databind.JsonNode;
import tools.jackson.databind.json.JsonMapper;

import java.net.URI;
import java.net.http.HttpClient;
import java.net.http.HttpRequest;
import java.net.http.HttpResponse;
import java.nio.charset.StandardCharsets;
import java.time.Duration;
import java.util.Base64;
import java.util.LinkedHashMap;
import java.util.Map;

/**
 * Vault Transit client for PII encryption/decryption. Mirrors Go internal/clients/vault_client.go.
 *
 * <p>Calls the Vault HTTP API directly (matching the Go service, which used direct HTTP calls as a
 * workaround for the vault-client-go library): POST {address}/v1/transit/encrypt/{key} with a
 * base64-encoded plaintext, and POST {address}/v1/transit/decrypt/{key} with the {@code vault:v1:...}
 * ciphertext (response plaintext is base64-decoded). When Vault is disabled the calls are bypassed
 * and the value is returned unchanged, exactly as in Go.
 */
@Component
public class VaultClient {

    private static final Logger log = LoggerFactory.getLogger(VaultClient.class);

    private final IndividualProperties.Vault config;
    private final boolean enabled;
    private final HttpClient httpClient;
    private final JsonMapper mapper;
    private final VaultAuth auth;

    public VaultClient(IndividualProperties props) {
        this.config = props.getVault();
        this.enabled = config.isEnabled();
        this.mapper = JsonMapper.builder().build();
        // Matches Go vault_client.go httpClient timeout of 10s.
        this.httpClient = HttpClient.newBuilder()
                .connectTimeout(Duration.ofSeconds(10))
                .build();
        if (!enabled) {
            this.auth = null;
            log.info("Vault disabled - encryption/decryption will be bypassed");
        } else {
            this.auth = new VaultAuth(config.getAddress(), config.getRoleId(), config.getSecretId(), httpClient, mapper);
            this.auth.login();
            log.info("Vault client initialized (AppRole) (address={})", config.getAddress());
        }
    }

    /** Vault address with any trailing slash removed, so transit paths don't produce a double slash. */
    private String baseAddress() {
        String addr = config.getAddress();
        if (addr == null) {
            return "";
        }
        int end = addr.length();
        while (end > 0 && addr.charAt(end - 1) == '/') {
            end--;
        }
        return addr.substring(0, end);
    }

    /** Encrypts plaintext via Vault Transit. Returns plaintext unchanged when disabled/empty/already-encrypted. */
    public String encrypt(String plaintext, String key) {
        if (!enabled) {
            return plaintext;
        }
        if (plaintext == null || plaintext.isEmpty()) {
            return "";
        }
        // Already encrypted (vault:v1: prefix) - return as-is.
        if (plaintext.startsWith("vault:v1:")) {
            return plaintext;
        }

        String encoded = Base64.getEncoder().encodeToString(plaintext.getBytes(StandardCharsets.UTF_8));
        Map<String, Object> requestBody = new LinkedHashMap<>();
        requestBody.put("plaintext", encoded);

        String url = baseAddress() + "/v1/transit/encrypt/" + key;
        JsonNode data = call(url, requestBody, "encrypt");

        JsonNode ciphertextNode = data.get("ciphertext");
        String ciphertext = ciphertextNode == null ? null : ciphertextNode.asString();
        if (ciphertext == null || ciphertext.isEmpty()) {
            throw new RuntimeException("empty ciphertext in response");
        }
        return ciphertext;
    }

    /** Decrypts ciphertext via Vault Transit. Returns ciphertext unchanged when disabled/empty/not-encrypted. */
    public String decrypt(String ciphertext, String key) {
        if (!enabled) {
            return ciphertext;
        }
        if (ciphertext == null || ciphertext.isEmpty()) {
            return "";
        }
        // Not encrypted (no vault:v1: prefix) - return as-is.
        if (!ciphertext.startsWith("vault:v1:")) {
            return ciphertext;
        }

        Map<String, Object> requestBody = new LinkedHashMap<>();
        requestBody.put("ciphertext", ciphertext);

        String url = baseAddress() + "/v1/transit/decrypt/" + key;
        JsonNode data = call(url, requestBody, "decrypt");

        JsonNode plaintextNode = data.get("plaintext");
        String plaintextB64 = plaintextNode == null ? null : plaintextNode.asString();
        if (plaintextB64 == null || plaintextB64.isEmpty()) {
            throw new RuntimeException("empty plaintext in response");
        }
        byte[] decoded;
        try {
            decoded = Base64.getDecoder().decode(plaintextB64);
        } catch (IllegalArgumentException e) {
            throw new RuntimeException("failed to decode plaintext: " + e.getMessage());
        }
        return new String(decoded, StandardCharsets.UTF_8);
    }

    private JsonNode call(String url, Map<String, Object> requestBody, String op) {
        String json;
        try {
            json = mapper.writeValueAsString(requestBody);
        } catch (Exception e) {
            throw new RuntimeException("failed to marshal request: " + e.getMessage());
        }

        HttpResponse<String> resp;
        try {
            auth.ensureToken();
            resp = send(url, json);
            // Token expired/revoked: re-login via AppRole and retry once.
            if (resp.statusCode() == 403) {
                auth.login();
                resp = send(url, json);
            }
        } catch (RuntimeException e) {
            throw e;
        } catch (Exception e) {
            throw new RuntimeException("failed to " + op + " data: " + e.getMessage());
        }

        if (resp.statusCode() != 200) {
            throw new RuntimeException("vault returned status " + resp.statusCode() + ": " + resp.body());
        }

        JsonNode root;
        try {
            root = mapper.readTree(resp.body());
        } catch (Exception e) {
            throw new RuntimeException("failed to decode response: " + e.getMessage());
        }
        JsonNode data = root.get("data");
        if (data == null || data.isNull()) {
            throw new RuntimeException("empty response from vault");
        }
        return data;
    }

    private HttpResponse<String> send(String url, String json) throws Exception {
        HttpRequest req = HttpRequest.newBuilder()
                .uri(URI.create(url))
                .timeout(Duration.ofSeconds(10))
                .header("Content-Type", "application/json")
                .header("X-Vault-Token", auth.token())
                .POST(HttpRequest.BodyPublishers.ofString(json))
                .build();
        return httpClient.send(req, HttpResponse.BodyHandlers.ofString());
    }
}

File: conformance/code/individual-go/internal/clients/vault_client.go
package clients

import (
	"bytes"
	"context"
	"encoding/base64"
	"encoding/json"
	"fmt"
	"io"
	"net/http"
	"strings"
	"time"

	"individual/internal/config"

	"github.com/rs/zerolog/log"
)

// VaultClient handles encryption/decryption via HashiCorp Vault.
// Key lifecycle (creation, rotation, config) is managed entirely by Vault;
// this client only performs encrypt/decrypt against the Transit engine.
type VaultClient interface {
	Encrypt(ctx context.Context, plaintext string, key string) (string, error)
	Decrypt(ctx context.Context, ciphertext string, key string) (string, error)
}

type vaultClient struct {
	httpClient *http.Client
	config     *config.VaultConfig
	auth       *vaultAuth
	enabled    bool
}

// NewVaultClient creates a new Vault client that authenticates via AppRole.
func NewVaultClient(cfg *config.VaultConfig) (VaultClient, error) {
	if !cfg.Enabled {
		log.Info().Msg("Vault disabled — encryption/decryption will be bypassed")
		return &vaultClient{enabled: false}, nil
	}

	httpClient := &http.Client{
		Timeout: 10 * time.Second,
	}

	auth := newVaultAuth(cfg.Address, cfg.RoleID, cfg.SecretID, httpClient)
	if err := auth.login(context.Background()); err != nil {
		return nil, fmt.Errorf("vault approle login failed: %w", err)
	}

	log.Info().Str("address", cfg.Address).Msg("Vault client initialized (AppRole)")

	return &vaultClient{
		httpClient: httpClient,
		config:     cfg,
		auth:       auth,
		enabled:    true,
	}, nil
}

// doAuthed issues an authenticated request to Vault. If the token has expired
// or been revoked (HTTP 403), it performs a fresh AppRole login and retries
// the request once.
func (v *vaultClient) doAuthed(ctx context.Context, method, url string, body []byte) (*http.Response, error) {
	if err := v.auth.ensureToken(ctx); err != nil {
		return nil, fmt.Errorf("vault token unavailable: %w", err)
	}
	attempt := func() (*http.Response, error) {
		var rdr io.Reader
		if body != nil {
			rdr = bytes.NewBuffer(body)
		}
		req, err := http.NewRequestWithContext(ctx, method, url, rdr)
		if err != nil {
			return nil, err
		}
		if body != nil {
			req.Header.Set("Content-Type", "application/json")
		}
		req.Header.Set("X-Vault-Token", v.auth.Token())
		return v.httpClient.Do(req)
	}

	resp, err := attempt()
	if err != nil {
		return nil, err
	}
	if resp.StatusCode == http.StatusForbidden {
		resp.Body.Close()
		if lerr := v.auth.login(ctx); lerr != nil {
			return nil, fmt.Errorf("vault re-login after 403 failed: %w", lerr)
		}
		return attempt()
	}
	return resp, nil
}

// Encrypt encrypts plaintext using the Vault Transit engine.
func (v *vaultClient) Encrypt(ctx context.Context, plaintext string, key string) (string, error) {
	if !v.enabled {
		return plaintext, nil
	}

	if plaintext == "" {
		return "", nil
	}

	// Check if already encrypted (vault:v1: prefix)
	if strings.HasPrefix(plaintext, "vault:v1:") {
		return plaintext, nil
	}

	// Base64 encode the plaintext as required by Vault Transit
	encoded := base64.StdEncoding.EncodeToString([]byte(plaintext))

	requestBody := map[string]interface{}{
		"plaintext": encoded,
	}

	jsonData, err := json.Marshal(requestBody)
	if err != nil {
		return "", fmt.Errorf("failed to marshal request: %w", err)
	}

	url := fmt.Sprintf("%s/v1/transit/encrypt/%s", strings.TrimRight(v.config.Address, "/"), key)
	resp, err := v.doAuthed(ctx, http.MethodPost, url, jsonData)
	if err != nil {
		return "", fmt.Errorf("failed to encrypt data: %w", err)
	}
	defer resp.Body.Close()

	if resp.StatusCode != http.StatusOK {
		body, _ := io.ReadAll(resp.Body)
		return "", fmt.Errorf("vault returned status %d: %s", resp.StatusCode, string(body))
	}

	var vaultResp struct {
		Data struct {
			Ciphertext string `json:"ciphertext"`
		} `json:"data"`
	}

	if err := json.NewDecoder(resp.Body).Decode(&vaultResp); err != nil {
		return "", fmt.Errorf("failed to decode response: %w", err)
	}

	if vaultResp.Data.Ciphertext == "" {
		return "", fmt.Errorf("empty ciphertext in response")
	}

	return vaultResp.Data.Ciphertext, nil
}

// Decrypt decrypts ciphertext using the Vault Transit engine.
func (v *vaultClient) Decrypt(ctx context.Context, ciphertext string, key string) (string, error) {
	if !v.enabled {
		return ciphertext, nil
	}

	if ciphertext == "" {
		return "", nil
	}

	// Check if encrypted (vault:v1: prefix)
	if !strings.HasPrefix(ciphertext, "vault:v1:") {
		return ciphertext, nil
	}

	requestBody := map[string]interface{}{
		"ciphertext": ciphertext,
	}

	jsonData, err := json.Marshal(requestBody)
	if err != nil {
		return "", fmt.Errorf("failed to marshal request: %w", err)
	}

	url := fmt.Sprintf("%s/v1/transit/decrypt/%s", strings.TrimRight(v.config.Address, "/"), key)
	resp, err := v.doAuthed(ctx, http.MethodPost, url, jsonData)
	if err != nil {
		return "", fmt.Errorf("failed to decrypt data: %w", err)
	}
	defer resp.Body.Close()

	if resp.StatusCode != http.StatusOK {
		body, _ := io.ReadAll(resp.Body)
		return "", fmt.Errorf("vault returned status %d: %s", resp.StatusCode, string(body))
	}

	var vaultResp struct {
		Data struct {
			Plaintext string `json:"plaintext"`
		} `json:"data"`
	}

	if err := json.NewDecoder(resp.Body).Decode(&vaultResp); err != nil {
		return "", fmt.Errorf("failed to decode response: %w", err)
	}

	if vaultResp.Data.Plaintext == "" {
		return "", fmt.Errorf("empty plaintext in response")
	}

	decoded, err := base64.StdEncoding.DecodeString(vaultResp.Data.Plaintext)
	if err != nil {
		return "", fmt.Errorf("failed to decode plaintext: %w", err)
	}

	return string(decoded), nil
}
